src/transformers/models/idefics/processing_idefics.py

```diff
@@ -345,39 +369,72 @@ def image_tokens(last_was_image):
+                    # Update the padded_image_tensor with the values from current_images
+                    indices = tf.reshape(tf.range(num_images), (-1, 1))
+                    updates = current_images
+                    padded_image_tensor = tf.tensor_scatter_nd_update(padded_image_tensor, indices, updates)
             else:
-                padded_image_tensor = torch.zeros(max_num_images, *self.default_image_dims)
+                if return_tensors == "pt":
+                    padded_image_tensor = torch.zeros(max_num_images, *self.default_image_dims)
+                elif return_tensors == "tf":
+                    padded_image_tensor = tf.zeros((max_num_images, *self.default_image_dims))
 
             output_images.append(padded_image_tensor)
-            output_input_ids.append(torch.tensor(padded_input_ids))
-            output_attention_masks.append(torch.tensor(attention_mask))
-
-        output_input_ids = torch.stack(output_input_ids)
-        output_images = torch.stack(output_images)
-        output_attention_masks = torch.stack(output_attention_masks)
+            if return_tensors == "pt":
+                output_input_ids.append(torch.tensor(padded_input_ids))
+                output_attention_masks.append(torch.tensor(attention_mask))
+            elif return_tensors == "tf":
+                output_input_ids.append(tf.convert_to_tensor(padded_input_ids, dtype=tf.int32))
+                output_attention_masks.append(attention_mask)
+
+        if return_tensors == "pt":
+            output_input_ids = torch.stack(output_input_ids)
+            output_images = torch.stack(output_images)
+            output_attention_masks = torch.stack(output_attention_masks)
+        elif return_tensors == "tf":
+            output_input_ids = tf.stack(output_input_ids)
+            output_images = tf.stack(output_images)
+            output_attention_masks = tf.stack(output_attention_masks)
 
         if at_least_one_image:
-            image_attention_mask, _ = image_attention_mask_for_packed_input_ids(output_input_ids, self.tokenizer)
+            image_attention_mask, _ = image_attention_mask_for_packed_input_ids(
+                output_input_ids, self.tokenizer, return_tensors
+            )
             image_attention_mask = incremental_to_binary_attention_mask(
-                image_attention_mask, num_classes=max_num_images
+                image_attention_mask, return_tensors, num_classes=max_num_images
             )
         else:
             # in full language mode we set the image mask to all-0s
-            image_attention_mask = torch.zeros(
-                output_input_ids.shape[0], output_input_ids.shape[1], 1, dtype=torch.bool
-            )
-
+            if return_tensors == "pt":
+                image_attention_mask = torch.zeros(
+                    output_input_ids.shape[0], output_input_ids.shape[1], 1, dtype=torch.bool
+                )
+            elif return_tensors == "tf":
```

Comment:
General comment: For `processing_idefics.py`, rather than blending the PT and TF code paths into the same function, it might be easier to just separate them. Keep the top-level function `image_attention_mask_for_packed_input_ids`, but it just calls either `image_attention_mask_for_packed_input_ids_pt` or `image_attention_mask_for_packed_input_ids_tf` based on the value of `return_tensors`. This way, you can keep the original Torch code intact (just move it to `image_attention_mask_for_packed_input_ids_pt`)

This makes this PR a bit more modular and easy to read/review/modify in future - and shouldn't make it much longer, since the PT and TF code call very different functions!

(Also, all the PyTorch people can't read TF and they'll get stressed if we put it in their functions...)

Reply:
Yup totally agree, it was very unreadable. I just did the refactor [here](https://github.com/huggingface/transformers/pull/26870/commits/752488fd112d0727ceca1e515ecfa3d819bd8e08) as you suggested, please take a quick look when you get a chance